### ISSUE TITLE
Investigate and experiment with the improved exp

### ISSUE THREAD

**Issue by KalaivaniMCW** [tenstorrent/tt-metal#24306]
- [ ] Implement the experimental versions of exp from https://github.com/tenstorrent/tt-llk/tree/nmaurice/exp-accurate to tt-metal for easier testing
 refer [link](https://github.com/tenstorrent/tt-llk/blob/nmaurice/exp-accurate/tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_exp.h#L28)

- [ ] share the branch/draft PR with LLK team for feedback

- [ ] Compare ULP for a specific input ranges, with plot to see the diff between existing & improved version

- [ ] extract `exp` outputs for the formula `pow(a,b) = exp( b * log(a))` and compare with torch fns
ex: 
1. input = 9, exponent = 2 , exp_input = 2 * log(9)
2. input/base = 10000, exponent = 1.7984,  exp_input = 1.7984 * log (10000)
I've picked these two inputs because we have open issues for these. Try other values as per your judgement.

**Comment by VirdhatchaniKN** [comment]
Updates on the accurate version of aproach

- Analysis report of expm1, selu, elu when tried using the updated version of exp : [Exp analysis.pdf](https://github.com/user-attachments/files/21028217/Exp.analysis.pdf)
- Pow gives register spill issues, but the inputs provided matches the expected. Below shown are some inputs tested with updated version of exp

<google-sheets-html-origin>
Input | Torch result | TT Result | status (Compare_equal)
-- | -- | -- | --
-88 | 6.06E-39 | 0.00E+00 | failed
-89 | "2.2041e-39" | "1.5428e-38" | failed
-90 | "8.2652e-40" | "4.2061e-38" | failed
-59 | 2.38E-26 | 2.37E-26 | failed
-58 | 6.46E-26 | 6.42E-26 | failed
-55 | 1.30E-24 | 1.30E-24 | pass
-30 | 9.37E-14 | 9.33E-14 | failed
-20 | "2.0664e-09" | "2.0518e-09" | failed
-15 | "3.0547e-07" | "3.0361e-07" | failed
-9 | 1.00E-04 | 1.00E-04 | pass
-10 | 4.53E-05 | 4.53E-05 | pass
0 | 1 | 1 | pass
0.5 | 1.6484 | 1.6484 | pass
1.7969 | 6.0312 | 6 | failed
2 | 7.375 | 7.375 | pass
4.4062 | 82 | 81.5 | failed
9 | 8096 | 8064 | failed
20 | 4.84E+08 | 4.84E+08 | pass
20.75 | "1.0276e+09" | "1.0234e+09" | failed
80 | "5.5493e+34" | "5.5168e+34" | failed
81 | "1.5058e+35" | "1.4993e+35" | failed
82 | "4.1019e+35" | "4.0760e+35" | failed
83 | 1.11E+36 | 1.11E+36 | pass
84 | "3.0323e+36" | "3.0115e+36" | failed
85 | "8.2246e+36" | "8.1831e+36" | failed
86 | "2.2431e+37" | "2.2265e+37" | failed
87 | "6.0812e+37" | "6.0480e+37" | failed
88 | "1.6549e+38" | "1.6482e+38" | failed
89 | inf | inf | pass
90 | inf | inf | pass
999424 | inf | inf | pass

**Comment by KalaivaniMCW** [comment]
<img width="856" alt="Image" src="https://github.com/user-attachments/assets/c0bc743a-b38c-4ce9-84df-6d938a83dc76" />


When testing the exp value corresponding to 9 ** 2 we get 81.5 (the existing version gives ~78)

<img width="567" alt="Image" src="https://github.com/user-attachments/assets/83ad6091-ed1c-47c5-9e31-e520866bb021" />

**Comment by VirdhatchaniKN** [comment]
Update : 
### Analysis for range [-100,100]

- **Report with graphs** : [Analysis.pdf](https://github.com/user-attachments/files/21123531/Analysis.pdf)

- **Profiling result :** The accurate version is significantly slower

<google-sheets-html-origin>
Input Shape | Main branch | Accurate branch | Overhead | % Slower
-- | -- | -- | -- | --
[1, 1, 32, 32] | 3,330 ns | 4,493 ns | 1,163 ns | 34.90%
[1, 1, 1024, 1024] | 37,981 ns | 58,930 ns | 20,949 ns | 55.20%

**Report summary :** 

- **ELU**:
  - ttnn.elu matches torch.elu perfectly.
  - No accuracy issues; clipping in exp has no effect here.
- **EXP**:
  - In range [-100, -75], ttnn.exp shows a clear mismatch (spikes, underflow).
  - For x > -75, both versions align well.
  - Clipping the input (e.g., to x ≥ -60) resolves the mismatch without affecting accuracy
- **SELU**:
  - Formula uses exp internally for negative inputs.
  - ttnn.selu closely matches torch.selu, but a tiny difference appears in range [75, 100] — likely due to bfloat16 precision.
  - Clipping exp input has no visible effect on SELU results.
- **EXPM1**:
  - ttnn.expm1 is accurate until x ≈ 85, but diverges after that.
  - For x > 85, graph grows in steps and flattens early (~2.6e38 vs torch > 3e38).
  - Clipping again shows no major impact on results.

**Comment by KalaivaniMCW** [comment]
Testing pow formula with activations, using exp21_f: 

```
ttnn.multiply(tt_ina, tt_inb, 
                input_tensor_a_activations=[ttnn.UnaryWithParam(ttnn.UnaryOpType.LOG)], 
                activations=[ttnn.UnaryWithParam(ttnn.UnaryOpType.EXP, False)], 
                use_legacy=False)
```
Inputs:
```
    torch_input_a = torch.tensor([9.0, 100000, 5.0], dtype= torch.float32)
    torch_input_b = torch.tensor([2.0, 1.7984, 3.0], dtype= torch.float32)
```

Intermediate results for exp( b * log(a))
```
tt logA -  ttnn.Tensor([ 2.19499, 11.50271,  1.60627], shape=Shape([3]), dtype=DataType::FLOAT32, layout=Layout::TILE)
torch logA -  tensor([ 2.1972, 11.5129,  1.6094])
tt log mul -  ttnn.Tensor([ 4.38999, 20.68647,  4.81881], shape=Shape([3]), dtype=DataType::FLOAT32, layout=Layout::TILE)
torch log mul -  tensor([ 4.3944, 20.7048,  4.8283])
tt exp -  ttnn.Tensor([80.42142, 963669248.00000, 123.58892], shape=Shape([3]), dtype=DataType::FLOAT32, layout=Layout::TILE)
torch exp -  tensor([8.1000e+01, 9.8175e+08, 1.2500e+02])
```

torch.pow vs ttnn pow (using ttnn.mul w pre & post activations): 

```
golden: tensor([8.1000e+01, 9.8175e+08, 1.2500e+02]) 
TTNN: tensor([8.0421e+01, 9.6367e+08, 1.2359e+02])
```


**For comparison with existing exp :**
```
tt logA -  ttnn.Tensor([ 2.19499, 11.50271,  1.60627], shape=Shape([3]), dtype=DataType::FLOAT32, layout=Layout::TILE)
torch logA -  tensor([ 2.1972, 11.5129,  1.6094])
tt log mul -  ttnn.Tensor([ 4.38999, 20.68647,  4.81881], shape=Shape([3]), dtype=DataType::FLOAT32, layout=Layout::TILE)
torch log mul -  tensor([ 4.3944, 20.7048,  4.8283])
tt exp -  ttnn.Tensor([77.93266, 911893696.00000, 121.14920], shape=Shape([3]), dtype=DataType::FLOAT32, layout=Layout::TILE)
torch exp -  tensor([8.1000e+01, 9.8175e+08, 1.2500e+02])
```

torch.pow vs ttnn.pow (existing)
```
golden: tensor([8.1000e+01, 9.8175e+08, 1.2500e+02]) 
TTNN: tensor([7.8192e+01, 9.2114e+08, 1.2236e+02])
```

Although I see an improvement, it doesn't match for all values.

cc: @nmauriceTT  @cmaryanTT

**Comment by nmauriceTT** [comment]
These are great & useful results.

I've looked in details at the pow results. 

Since any error in the logarithm function is going to be magnified by the exponential function, I've computed both the final relative error, as well as the relative error of each step 
(i.e. what would be the final error after exp if ignore errors in the logarithm).


 . | . |   |   |  | | ttnn.log(a)   |   |  | b * ttnn.log(a)  |   |  | ttnn.exp_21f(b * ttnn.log(a))  |   | |  ttnn.old.exp(b * ttnn.log(a))  |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
 **a** | **b** | **torch.log(a)**  | **b * torch.log(a)**  | **golden**  | **value** | **added error** | **total error** | **value** | **added error** | **total error** | **value** | **added error** | **total error** | **value** | **error** | **total error**
9 | 2 | 2.1972 | 4.3944 | 81 | 2.19499 | 0.0010 | 0.0010 | 4.38999 | 0.0000 | 0.0010 | 77.93266 | N/A | 0.0379 | 80.42142 | 0.0027 | 0.0071
100000 | 1.7984 | 11.5129 | 20.7048 | 9.82E+08 | 11.50271 | 0.0009 | 0.0009 | 20.68647 | 0.0000 | 0.0009 | 911893696 | N/A | 0.0712 | 963669248 | 0.0002 | 0.0184
5 | 3 | 1.6094 | 4.8283 | 125 | 1.60627 | 0.0019 | 0.0019 | 4.81881 | 0.0000 | 0.0020 | 121.1492 | N/A | 0.0308 | 123.58892 | 0.0018 | 0.0113

[Spreadsheet Version](https://github.com/user-attachments/files/21124846/Accuracy.-.exp_21f.ods)

In this case, the relative error rate of `exp_21f` is < 0.5% if we exclude the error of the logarithm. On the other hand, the relative error rate of `ttnn.log` is `< 0.2%`.

But what is interesting is that small errors in the logarithm will have a large influence over the final results. 
For instance, for `a=9, b=2`, `PerfectExp( b * ttnn.log(a)) = 80.64` 
i.e. making `ttnn.exp` 'perfect' *should* further reduce the error by a factor of 1.6 (relative error `0.71% -> 0.44%`) , but making `ttnn.log` perfect would reduce it by a factor of 2.6 (relative error `0.71% -> 0.27%`)

**Comment by cmaryanTT** [comment]
@nmauriceTT - that's a really good insight

**Comment by VirdhatchaniKN** [comment]
Update on exp :

- Attached is the comparison of ops from main branch and accurate version of exp with ULP graphs : [Exp analysis - Google Docs.pdf](https://github.com/user-attachments/files/21161422/Exp.analysis.-.Google.Docs.pdf)
- CSV Comparison : Generated CSV for torch vs TTNN, abs difference, ULP.
  - Attached are my CSV results
    - [ttnn.elu_bf16__1.0_results.csv](https://github.com/user-attachments/files/21161437/ttnn.elu_bf16__1.0_results.csv)
    - [ttnn.exp_bf16__noscalar_results.csv](https://github.com/user-attachments/files/21161434/ttnn.exp_bf16__noscalar_results.csv)
    - [ttnn.expm1_bf16__noscalar_results.csv](https://github.com/user-attachments/files/21161435/ttnn.expm1_bf16__noscalar_results.csv)
    - [ttnn.selu_bf16__noscalar_results.csv](https://github.com/user-attachments/files/21161436/ttnn.selu_bf16__noscalar_results.csv)
  - Analysis
    - Selu, elu works fine
    - EXP
      - Accurate for x ≥ -87.
      - For x < -87, ULP Error is high because TTNN is clipped to zero, but torch value is around zero
    - EXPM1
      - Matches torch until x = 85.
      - Beyond that: grows in steps, flattens early at ~2.6e38. Inaccurate for x > 85, especially [85, 100].

**Comment by vmilicevicTT** [comment]
@cmaryanTT , can you please help prioritize this?  P1/P2?  Thanks

**Comment by cmaryanTT** [comment]
@vmilicevicTT - Added P1 prioritization, but no LLK team involvement needed at this time. We have an implementation and are working through the integration issues.

**Comment by VirdhatchaniKN** [comment]
Update on exp : 

- PR : https://github.com/tenstorrent/tt-metal/pull/24994
- Pipelines triggered after updating exp2 implementation. Will update once its completed
- Working on MLA Prefill failure
  - The test runs for two inputs, one of bf16 and other of bf8_b data type
  - On using DPRINT, The input to `sub_exp_block` (Tile with exp causing failure) is within range -5 to 5.
    - BF16 Passes in this range with ULP 2
    - BF8 results with ULP, Result graphs : [bf8 results.pdf](https://github.com/user-attachments/files/21369106/bf8.results.pdf)
  - Working on `ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/sdpa.cpp` to get the entire input and output tensor of exp tile

**Comment by VirdhatchaniKN** [comment]
**Update** : 

- PR : https://github.com/tenstorrent/tt-metal/pull/25946
- Max ULP Set in test : 1
- Working range of Exp2 : [-87,88.5]
  - Overflow from 88.5(inf)
  - Underflow till -87(<0)
- ULP Results for working range

<google-sheets-html-origin>
Low | High | Step | Max ULP Delta
-- | -- | -- | --
-90 | -87.1 | 0.01 | 109 (Mismatching values mentioned below graphs)
-87 | 88.6 | 0.01 | 1

  - Below attached is the Updated ULP Graph for BF16 Range from https://github.com/tenstorrent/tt-metal/pull/24143:
    - The spike in the graph is when input < -87 (ULP = 109)
    - The ULP error graph stops when input > 88.5 because result overflows to inf.

### ULP Plots
Note : ULP 109 is expected to be with values lesser or equal to -87, but graph shows different because the script does batch testing and then plots the first value of each batch against that batch’s max_ulp_error.
- ULP Plot for full range input

<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/27ccebf9-8c31-4d71-872b-297ce73d48cd" />

- Clamped ULP Graph for full range input

<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/f1862468-da98-4289-8962-284d0e2ff34e" />

- Scatter plot for full range input

<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/cf2be9e6-e4b5-4806-8bb4-db2d8361157d" />

- Mismatching values :
  - Input tensor value: -87.5
    - Calculated tensor value: 0.0
    - Expected tensor value: 1.0010069081221042e-38
  - Input tensor value: -88.0
    - Calculated tensor value: 0.0
    - Expected tensor value: 6.06114274642742e-39
  - Input tensor value: -88.5
    - Calculated tensor value: 0.0
    - Expected tensor value: 3.6734198463196485e-39
  - Input tensor value: -89.0
    - Calculated tensor value: 0.0
    - Expected tensor value: 2.204051907791789e-39
  - Input tensor value: -89.5
    - Calculated tensor value: 0.0
    - Expected tensor value: 1.3775324423698682e-39
  - Input tensor value: -90.0
    - Calculated tensor value: 0.0
    - Expected tensor value: 8.265194654219209e-40

**Comment by VirdhatchaniKN** [comment]
Update : 

- PR : https://github.com/tenstorrent/tt-metal/pull/25946
- Max ULP Set in test : 1
- Working range of Exp2 : [-87,88.5]
  - Overflow from 88.5(inf)
  - Underflow till -87(<0)
- Pipeline update : In Progress
- ULP Graphs for full range with main vs 21F branch : 
  - Detailed report of comparison with main : [Exp_ULP_Graphs.pdf](https://github.com/user-attachments/files/21711094/Exp_ULP_Graphs.pdf)
  - Full range comparison with main

<img width="902" height="664" alt="Image" src="https://github.com/user-attachments/assets/8518d415-13d3-4461-82d8-c998a0aa3c6f" />

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/pull/25946 - Exp branch merged to main

**Comment by cmaryanTT** [comment]
@VirdhatchaniKN - can you please close this issue, and open a new clean issue for tracking re-evaluating all exp-dependent functions, to see what their accuracy is right now, and whether they should by default use the new exp as their method of implementation.

**Comment by VirdhatchaniKN** [comment]
> [@VirdhatchaniKN](https://github.com/VirdhatchaniKN) - can you please close this issue, and open a new clean issue for tracking re-evaluating all exp-dependent functions, to see what their accuracy is right now, and whether they should by default use the new exp as their method of implementation.

Sure,
New issue created for tracking re-evaluation of exp-dependent ops : https://github.com/tenstorrent/tt-metal/issues/26993
Hence, closing issue.
